pkg/bff/db/announcements.go

```diff
@@ -0,0 +1,215 @@
+)
+
+var (
+	lastUpdatedKey = []byte("last_updated")
+)
+
+// The Announcements type exposes methods for interacting with Network Announcements
+// in the database. Announcements are stored as compact, compressed JSON to minimize the
+// network requests and reduce storage requirements. This struct performs all necessary
+// serialization on the announcements before storing and retrieving the model object. The
+// announcement keys are ksuids - timestamp ordered unique IDs so that it is easy to scan
+// the trtl database to find the most recent announcements.
+//
+// Announcements implements the Collection interface
+type Announcements struct {
+	db        *DB
+	namespace string
+}
+
+// Ensure that Announcements implements the Collection interface.
+var _ Collection = &Announcements{}
+
+// Announcements constructs the collection type for interactions with network
+// announcement objects. This method is intended to be used with chaining, e.g. as
+// db.Announcements().Recent(), so to reduce the number of allocations a singleton
+// intermediate struct is used. Method calls to the collection are thread-safe.
+func (db *DB) Announcements() *Announcements {
+	db.muMakeAC.Do(func() {
+		db.announcements = &Announcements{
+			db:        db,
+			namespace: NamespaceAnnouncements,
+		}
+	})
+	return db.announcements
+}
+
+// Recent returns the set of results whose post date is after the not before timestamp,
+// limited to the maximum number of results. Last updated returns the timestamp that
+// any announcement was added or changed.
+func (a *Announcements) Recent(ctx context.Context, maxResults int, notBefore time.Time) (out *api.AnnouncementsReply, err error) {
+	out = &api.AnnouncementsReply{
+		Announcements: make([]*api.Announcement, 0, maxResults),
+	}
+
+	// Get the last updated value from the index key in the collection.
+	var value []byte
+	if value, err = a.db.Get(ctx, lastUpdatedKey, a.namespace); err != nil && !errors.Is(err, ErrNotFound) {
+		return nil, fmt.Errorf("could not fetch last updated: %s", err)
+	}
+
+	if len(value) > 0 {
+		if err = json.Unmarshal(value, &out.LastUpdated); err != nil {
+			return nil, fmt.Errorf("could not unmarshal last updated: %s", err)
+		}
+	}
+
+	var cursor trtl.Trtl_CursorClient
+	if cursor, err = a.db.trtl.Cursor(ctx, &trtl.CursorRequest{Namespace: a.namespace}); err != nil {
+		return nil, fmt.Errorf("could not connect to trtl cursor: %s", err)
+	}
+	defer cursor.CloseSend()
+
+	// Consume the cursor
+	// TODO: this filtering method is NOT ideal since it scans in ascending time order
+	// rather than reverse time order (trtl doesn't support reverse ordering). This
+	// means as the number of announcements in the collection grows, this method will
+	// get slower and slower. Worse, we're loading everything in memory so we might
+	// overload the pod. We need to either figure out a different key-value storage
+	// and ordering mechanism or implement reverse seek in trtl.
```

Comment:
Agreed this solution feels a little clunky and relies on some implicit behavior with the ksuids and trtl. Maybe we need something like an "ordered" namespace defined in trtl? Or if we don't care about keeping the older announcements around, make the records ephemeral somehow so that we only store the last month of entries?

Reply:
Technically the namespace is ordered, it's just ordered ascending and we need it ordered descending. If there was a ksuid that was naturally descending that would be ideal, but that's pretty difficult to do. You can think of it as us having an array of posts that are sorted in time order and we want to access the last n posts. 

In terms of ephemerality, we don't really want to lose announcements - it's good to have a record of them. 

One thing I thought of was to have per month collections. e.g. the key would be 2022-06 and then all the announcements for that month would be there. It would reduce the speed of writes (it would always be 2-3 gets and a put) and reading the entire collection of announcements would take longer, but it might solve the problem because we wouldn't have to seek.